### PR TITLE
Use pytest fixtures in functional tests.

### PR DESCRIPTION
**Pull request by plietar**
Usign fixtures provides clarity in tests, and gives us more flexibility. This will allow us to dynamically provide different implementations of the fixtures based on command line arguments, such as starting our own cchost process.

### CODE REVIEW

File: test/infra/fixtures.py [hunk 1]
```diff
@@ -2,79 +2,105 @@
...
+    Create a Client instance to connect to the test SCITT service.
+
+    Most tests will want to use the `client` fixture instead, which resets the
+    service to a known good state.
+    """
+    return Client(service_url, development=True, member_auth=member_auth)
+
+
+@pytest.fixture(scope="class")
+def configure_service(base_client: Client):
```
Comment (letmaik) [inline test/infra/fixtures.py:50]: I think if we rely heavily on fixtures then we should use types more often and check the types since the IDE or mypy won't be able to do it for us.
I found https://github.com/untitaker/pytest-fixture-typecheck which uses https://github.com/agronholm/typeguard. The code of pytest-fixture-typecheck is trivial enough (4 lines of code) to be vendored into our `conftest.py`.

File: test/infra/fixtures.py [hunk 2]
```diff
@@ -2,79 +2,105 @@
...
-        self.x5c_ca = x5c_ca
 
-    def __enter__(self):
-        self.did_web_server.__enter__()
+@pytest.fixture(scope="session")
+def member_auth(member_auth_path):
+    return (member_auth_path[0].read_text(), member_auth_path[1].read_text())
+
+
+@pytest.fixture(scope="class")
```
[the remark this answers is not in the document]
Comment (plietar) [reply test/infra/fixtures.py:38]: It's fine to have tests that aren't in a class, `pytest` will treat them as if they each had their own class around them.

In those cases, class and function is almost the same thing with one exception, that multiple parametrization of a single test will share a single class scope but each have their own function scope.

Making this `class` gives a bit of flexibility over `function`, since we can group tests together and let them share a client / did server / ... if we want to.

File: test/test_cli.py [hunk 3]
```diff
@@ -21,35 +21,40 @@
...
-    url = os.environ.get("CCF_URL", "https://127.0.0.1:8000")
-    server_args = {
-        "url": url,
-        "development": True,
-    }
+@pytest.fixture
+def run(request):
+    def f(*cmd, _with_service_url=False, _with_member_auth=False, **kwargs):
+        if _with_service_url:
+            kwargs["url"] = request.getfixturevalue("service_url")
```
[the remark this answers is not in the document]
Comment (plietar) [reply test/test_cli.py:28]: If you use #21 and have the test framework start cchost, but you only run tests which don't need a service, then putting them as fixture arguments will start the service regardless. Using `getfixturevalue` will request the fixture (and start the server) lazily, only if it is actually needed by a test you selected.

This is admittedly a corner case and not super essential to handle, so I'm kind of ambivalent about it.

File: test/test_cli.py [hunk 4]
```diff
@@ -21,35 +21,40 @@
...
-
-def test_smoke_test(tmp_path: Path):
-    url = os.environ.get("CCF_URL", "https://127.0.0.1:8000")
-    server_args = {
-        "url": url,
-        "development": True,
-    }
+@pytest.fixture
+def run(request):
+    def f(*cmd, _with_service_url=False, _with_member_auth=False, **kwargs):
```
[the remark this answers is not in the document]
Comment (plietar) [reply test/test_cli.py:26]: Yeah, this probably makes sense to do.

File: test/test_cli.py [hunk 5]
```diff
@@ -21,35 +21,40 @@
...
 
+def test_smoke_test(run, client, tmp_path: Path):
     trust_store = tmp_path / "store"
     trust_store.mkdir()
-    (trust_store / "service.json").write_text(
-        json.dumps(Client(**server_args).get_parameters())
-    )
+    (trust_store / "service.json").write_text(json.dumps(client.get_parameters()))
 
     (tmp_path / "claims.json").write_text(json.dumps({"foo": "bar"}))
```
[the remark this answers is not in the document]
Comment (plietar) [reply test/test_cli.py:60]: I actually want to test the `propose_ca_certs` command which is why I'm using the class directly instead of the fixture, as the latter does the CA cert stuff automatically.

In general I think this module should avoid using the python APIs as much as possible (yeah the `client.get_parameters()` call above is not great, but also making a CLI for just this seems overkill, maybe).

File: test/test_cli.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (plietar) [reply test/test_cli.py:28]: I'll add a comment to explain this.

File: test/test_cli.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (plietar) [reply test/test_cli.py:60]: I'll add a comment to explain this, similar to `test_default_did_port`.

File: test/test_perf.py [hunk 6]
```diff
@@ -1,66 +0,0 @@
-# Copyright (c) Microsoft Corporation.
```
[the remark this answers is not in the document]
Comment (plietar) [reply test/test_perf.py:1]: Well the move to fixtures breaks this, and since we never run it anyway it's probably already broken, so seems like I might as well remove it.

If we care about it we should fix it and run it in CI.

File: test/test_auth.py [hunk 7]
```diff
@@ -5,19 +5,35 @@
...
-        claims = crypto.sign_json_claimset(identity, {"foo": "bar"})
+class TestAuthentication:
+    def not_allowed(self, f):
+        with pytest.raises(ServiceError, match="InvalidAuthenticationInfo"):
+            f()
 
-        def setup(*, allow_unauthenticated, required_claims=None):
-            fixture.configure_service(
+    @pytest.fixture(scope="class")
+    def claims(self, client, did_web):
```
Comment (letmaik) [inline test/test_auth.py:19]:
```suggestion
    def claims(self, did_web):
```

File: test/infra/fixtures.py (hunk 1 above)
Comment (plietar) [reply test/infra/fixtures.py:50]: I think I'll add the integration with `typeguard` in a separate PR.

**Review by letmaik — APPROVED** [review]
Some docs on how to write tests using fixtures would be useful, but that can be in a follow-up PR.

**Review by darracott — APPROVED** [review]
